### PR TITLE
World generation.

### PR DESCRIPTION
**Pull request by lunar-crater-ex**
## About the PR
A PR to assert my dominance as the first person to write over 1k+LoC (2.7kloc to be precise, not including the library)  for this codebase.
Includes worldgen MVP, biomes and wrecks for another PR.

**Media**
image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame
- [ ] This PR does not require an ingame showcase

**Changelog**
no changelog bot no fun

### CODE REVIEW

File: Content.Server/00Citadel/Worldgen/Components/Debris/SimpleDebrisSelectorComponent.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+/// </summary>
+[RegisterComponent]
+public sealed class SimpleDebrisSelectorComponent : Component
+{
+    [DataField("debrisTable", required: true)]
+    private List<EntitySpawnEntry> _entries = default!;
+
+    private EntitySpawnCollectionCache? _cache;
+
+    public EntitySpawnCollectionCache CachedDebrisTable
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Components/Debris/SimpleDebrisSelectorComponent.cs]: Comments?

File: Content.Server/00Citadel/Worldgen/Components/Debris/SimpleFloorPlanPopulatorComponent.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+[RegisterComponent]
+public sealed class SimpleFloorPlanPopulatorComponent : Component
+{
+    [DataField("entries", required: true, customTypeSerializer:typeof(PrototypeIdDictionarySerializer<List<EntitySpawnEntry>, ContentTileDefinition>))]
+    private Dictionary<string, List<EntitySpawnEntry>> _entries = default!;
+
+    private Dictionary<string, EntitySpawnCollectionCache>? _caches = null;
+
+    [ViewVariables]
+    public Dictionary<string, EntitySpawnCollectionCache> Caches
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Components/Debris/SimpleFloorPlanPopulatorComponent.cs]: Comment?

File: Content.Server/00Citadel/Worldgen/GridPointsNearEnumerator.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics.Contracts;
+
+namespace Content.Server._00Citadel.Worldgen;
+
+public struct GridPointsNearEnumerator
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/GridPointsNearEnumerator.cs]: shouldnt this have a comment?

File: Content.Server/00Citadel/Worldgen/Prototypes/BiomePrototype.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype.Array;
+
+namespace Content.Server._00Citadel.Worldgen.Prototypes;
+
+/// <summary>
+/// This is a prototype for...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Prototypes/BiomePrototype.cs]: this comment doesnt look finished

File: Content.Server/00Citadel/Worldgen/Prototypes/NoiseChannelPrototype.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    public NoiseCoordinateProcess? NoiseCoordinateProcess { get; } = null;
+
+    /// <summary>
+    /// The "center" of the range of values. Or the minimum if mapped 0 through 1.
+    /// </summary>
+    [DataField("minimum")]
+    public float Minimum { get; } = 0;
+}
+
+[Prototype("noiseChannel")]
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Prototypes/NoiseChannelPrototype.cs]: Comments?

File: Content.Server/00Citadel/Worldgen/Prototypes/NoiseChannelPrototype.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    public string[]? Parents { get; }
+
+    /// <inheritdoc/>
+    [NeverPushInheritance]
+    [AbstractDataField]
+    public bool Abstract { get; }
+
+}
+
+public struct NoiseGenerator
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Prototypes/NoiseChannelPrototype.cs]: Where is the comment?

File: Content.Server/00Citadel/Worldgen/Systems/BaseWorldSystem.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+﻿using Content.Server._00Citadel.Worldgen.Components;
+using JetBrains.Annotations;
+using Robust.Shared.Map;
+
+namespace Content.Server._00Citadel.Worldgen.Systems;
+
+/// <summary>
+/// This handles...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/BaseWorldSystem.cs]: what does it handle tho?

File: Content.Server/00Citadel/Worldgen/Prototypes/WorldgenConfigPrototype.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+﻿using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+
+namespace Content.Server._00Citadel.Worldgen.Prototypes;
+
+/// <summary>
+/// This is a prototype for...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Prototypes/WorldgenConfigPrototype.cs]: Hell yeah more ... protoypes

File: Content.Server/00Citadel/Worldgen/Systems/Biomes/BiomeSelectionSystem.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
+﻿using System.Linq;
+using Content.Server._00Citadel.Worldgen.Components;
+using Content.Server._00Citadel.Worldgen.Prototypes;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+
+namespace Content.Server._00Citadel.Worldgen.Systems.Biomes;
+
+/// <summary>
+/// This handles...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/Biomes/BiomeSelectionSystem.cs]: ... seems to be pretty popular these days

File: Content.Server/00Citadel/Worldgen/Systems/Debris/DebrisFeaturePlacerSystem.cs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+using Content.Server._00Citadel.Worldgen.Systems.GC;
+using Content.Server._00Citadel.Worldgen.Tools;
+using Robust.Shared.Map;
+using Robust.Shared.Map.Components;
+using Robust.Shared.Random;
+
+namespace Content.Server._00Citadel.Worldgen.Systems.Debris;
+
+/// <summary>
+/// This handles...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/Debris/DebrisFeaturePlacerSystem.cs]: another ... handler

File: Content.Server/00Citadel/Worldgen/Systems/DeferredSpawnSystem.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
+using Robust.Shared.Map;
+using Robust.Shared.Timing;
+
+namespace Content.Server._00Citadel.Worldgen.Systems;
+
+public sealed class DeferredSpawnSystem : EntitySystem
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/DeferredSpawnSystem.cs]: Where is my ... comment?

File: Content.Server/00Citadel/Worldgen/Systems/GC/GCQueueSystem.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+using Content.Server._00Citadel.Worldgen.Prototypes;
+using Robust.Shared.Configuration;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Robust.Shared.Timing;
+
+namespace Content.Server._00Citadel.Worldgen.Systems.GC;
+
+/// <summary>
+/// This handles...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/GC/GCQueueSystem.cs]: `Here is my ... comment

File: Content.Server/00Citadel/Worldgen/Systems/NoiseIndexSystem.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+/// <summary>
+/// This handles the noise index.
+/// </summary>
+public sealed class NoiseIndexSystem : EntitySystem
+{
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IPrototypeManager _prototype = default!;
+
+    public NoiseGenerator Get(EntityUid holder, string protoId)
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/NoiseIndexSystem.cs]: comment please

File: Content.Server/00Citadel/Worldgen/Systems/NoiseIndexSystem.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        else
+        {
+            var proto = _prototype.Index<NoiseChannelPrototype>(protoId);
+            var gen = new NoiseGenerator(proto, _random.Next());
+            idx.Generators[protoId] = gen;
+            return gen;
+        }
+    }
+
+    public float Evaluate(EntityUid holder, string protoId, Vector2 coords)
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/NoiseIndexSystem.cs]: what is this supposed to do?

File: Content.Server/00Citadel/Worldgen/Systems/WorldControllerSystem.cs [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+            return;
+        var ev = new WorldChunkUnloadedEvent(uid, chunk.Coordinates);
+        RaiseLocalEvent(chunk.Map, ref ev);
+        RaiseLocalEvent(uid, ref ev);
+        //_sawmill.Debug($"Unloaded chunk {ToPrettyString(uid)} at {coords}");
+    }
+
+    private const int PlayerLoadRadius = 2;
+
+    //TODO: This should be time-shared by an MC.
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/WorldControllerSystem.cs]: TODO

File: Content.Server/00Citadel/Worldgen/Systems/WorldControllerSystem.cs [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        if (count > 0)
+        {
+            var timeSpan = _gameTiming.RealTime - startTime;
+            _sawmill.Debug($"Loaded {count} chunks in {timeSpan.TotalMilliseconds:N2}ms.");
+        }
+    }
+
+    public EntityUid? GetOrCreateChunk(Vector2i chunk, EntityUid map, WorldControllerComponent? controller = null)
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/WorldControllerSystem.cs]: what we doing here? (no comment >:(

File: Content.Server/00Citadel/Worldgen/Systems/WorldgenConfigSystem.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+using Robust.Shared.Configuration;
+using Robust.Shared.Map;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+using Robust.Shared.Utility;
+
+namespace Content.Server._00Citadel.Worldgen.Systems;
+
+/// <summary>
+/// This handles...
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Systems/WorldgenConfigSystem.cs]: ...

File: Content.Server/00Citadel/Worldgen/Tools/EntitySpawnCollectionCache.cs [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        [ViewVariables]
+        public List<EntitySpawnEntry> Entries { get; set; } = new();
+        [ViewVariables]
+        public float CumulativeProbability { get; set; }
+    }
+
+    [ViewVariables]
+    private readonly Dictionary<string, OrGroup> _orGroups = new();
+
+    public EntitySpawnCollectionCache(IEnumerable<EntitySpawnEntry> entries)
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Tools/EntitySpawnCollectionCache.cs]: Commentary game is really weak with this one

File: Content.Server/00Citadel/Worldgen/Tools/FastNoise.cs [hunk 19]
```diff
@@ -0,0 +1,2970 @@
...
+#if FN_USE_DOUBLES
+using FN_DECIMAL = System.Double;
+#else
+using FN_DECIMAL = System.Single;
+#endif
+using System.Runtime.CompilerServices;
+
+namespace Content.Server._00Citadel.Worldgen.Tools;
+
+public class FastNoise
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Tools/FastNoise.cs]: Yo, lets write a giant class and not tell anyone whjat its supposed to handle

File: Content.Server/00Citadel/Worldgen/Tools/PoissonDiskSampler.cs [hunk 20]
```diff
@@ -0,0 +1,165 @@
+﻿using Robust.Shared.Random;
+
+namespace Content.Server._00Citadel.Worldgen.Tools;
+
+public sealed class PoissonDiskSampler
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/Tools/PoissonDiskSampler.cs]: No comments WOOO

File: Content.Server/00Citadel/Worldgen/WorldGen.cs [hunk 21]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Diagnostics.Contracts;
+
+namespace Content.Server._00Citadel.Worldgen;
+
+public static class WorldGen
```
Comment (TheLordME) [inline Content.Server/00Citadel/Worldgen/WorldGen.cs]: More no comments! Amazing

**Review by TheLordME — APPROVED** [review]
Moony asked for this

File: Content.Server/00Citadel/Worldgen/Tools/FastNoise.cs (hunk 19 above)
Comment (lunar-crater-ex) [reply Content.Server/00Citadel/Worldgen/Tools/FastNoise.cs]: This one doesn't belong to me, blame the original author
:HECK:

**Comment by lunar-crater-ex** [comment]
a wild spurious test failure :guh:

**Pull request merged**